### PR TITLE
8334502: gtest/GTestWrapper.java fails on armhf due to LogDecorations.iso8601_utctime_test

### PR DESCRIPTION
**Pull request by vpa1977**
Due to time_t transition time_t is now 64bit on Ubuntu. zone_hours and zone_min are printed using %02d specifier. This causes random value to be printed as the timezone offset.

Testing - armhf gtest passes on Ubuntu Oracular:
```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   gtest:all/server                                   1033  1033     0     0   
==============================
TEST SUCCESS
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334502](https://bugs.openjdk.org/browse/JDK-8334502): gtest/GTestWrapper.java fails on armhf due to LogDecorations.iso8601_utctime_test (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19782/head:pull/19782` \
`$ git checkout pull/19782`

Update a local copy of the PR: \
`$ git checkout pull/19782` \
`$ git pull https://git.openjdk.org/jdk.git pull/19782/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19782`

View PR using the GUI difftool: \
`$ git pr show -t 19782`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19782.diff">https://git.openjdk.org/jdk/pull/19782.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19782#issuecomment-2177901202)

### CODE REVIEW

File: src/hotspot/share/runtime/os.cpp [hunk 1]
```diff
@@ -213,8 +213,8 @@ char* os::iso8601_time(jlong milliseconds_since_19700101, char* buffer, size_t b
     abs_local_to_UTC = -(abs_local_to_UTC);
   }
   // Convert time zone offset seconds to hours and minutes.
-  const time_t zone_hours = (abs_local_to_UTC / seconds_per_hour);
-  const time_t zone_min =
+  const int zone_hours = (abs_local_to_UTC / seconds_per_hour);
+  const int zone_min =
     ((abs_local_to_UTC % seconds_per_hour) / seconds_per_minute);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/os.cpp:218]: I think you need to cast the expression to `int` to ensure the compiler won't complain, regardless of how `time_t` may actually be declared.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks

**Comment by dholmes-ora** [comment]
@vpa1977 hotspot changes require two reviewers before integration. Thanks

**Comment by vpa1977** [comment]
> @vpa1977 hotspot changes require two reviewers before integration. Thanks

Hi, 

@dholmes-ora I apologise for the mistake - the PR header had only 1 reviewer. I have tried to remove sponsor label, but that is not a valid label. Could you please advice how to proceed?

Best Regards, 
 Vladimir.

**Review by shipilev — APPROVED** [review]
I was initially concerned we are narrowing to `int` to please the format specifiers. Normally, I would have expected us to work out a format specifier based on `time_t` size, and put it in `globalDefinitions.hpp`. But this looks okay as the unblocking fix.

**Comment by shipilev** [comment]
> @dholmes-ora I apologise for the mistake - the PR header had only 1 reviewer. I have tried to remove sponsor label, but that is not a valid label. Could you please advice how to proceed?

My review is the second one, so we can proceed. Normally, you just wait for more reviewers, even if you have said `/integrate`.

**Comment by vpa1977** [comment]
> I was initially concerned we are narrowing to `int` to please the format specifiers. Normally, I would have expected us to work out a format specifier based on `time_t` size, and put it in `globalDefinitions.hpp`. But this looks okay as the unblocking fix.

Hmm, actually [`long        tm_gmtoff;`](https://github.com/vpa1977/jdk/blob/ff97727f9a7141451dcc9a2565d277368c4b5d99/src/hotspot/share/runtime/os.cpp#L183) is getting assigned to time_t and in the alternate branch we are doing a [static cast](https://github.com/vpa1977/jdk/blob/ff97727f9a7141451dcc9a2565d277368c4b5d99/src/hotspot/share/runtime/os.cpp#L187C6-L187C7). I guess the correct fix should've been changing [UTC_to_local](https://github.com/vpa1977/jdk/blob/ff97727f9a7141451dcc9a2565d277368c4b5d99/src/hotspot/share/runtime/os.cpp#L180) to long.

**Comment by dholmes-ora** [comment]
> I was initially concerned we are narrowing to `int` to please the format specifiers. Normally, I would have expected us to work out a format specifier based on `time_t` size, and put it in `globalDefinitions.hpp`. But this looks okay as the unblocking fix.

Those two variables, like many others in that code are very obviously just small integer values. I think `time_t` use should be minimised as much as possible e.g. `seconds_per_minute` et al. should really just be int values IMO.

> correct fix should've been changing [UTC_to_local](https://github.com/vpa1977/jdk/blob/ff97727f9a7141451dcc9a2565d277368c4b5d99/src/hotspot/share/runtime/os.cpp#L180) to long

If you do that won't it just introduce different conversion issues later in the code. This whole code is really quite a mess with regards to types, but that is partly because the API and structures used are also a mess when it comes to types.

**Comment by vpa1977** [comment]
> If you do that won't it just introduce different conversion issues later in the code. This whole code is really quite a mess with regards to types, but that is partly because the API and structures used are also a mess when it comes to types.


We might get something like this https://github.com/openjdk/jdk/commit/63005eaea5ddb5eee5d826433733a57741212af5 - removing both of static casts that I have introduced, and trading a static cast in Windows branch for a static cast in AIX one.
But I am not sure that I should introduce a change that touches the architecture that I can not test.

**Comment by dholmes-ora** [comment]
Note that in general we do not use `long` in shared code precisely because its size is platform dependent.

**Comment by shipilev** [comment]
Sorry, I did not mean that a follow-up work would be needed. I agree the cast is fine. It would probably be a bit less confusing to me to cast right in `printf` argument block, so it would be obvious no calculations are affected, and we are really only catering for the format specifiers. But this is a minor thing that does not deserve a follow-up, really.

**Pull request closed without merging**